plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/TransportReindexAction.java

```diff
@@ -0,0 +1,230 @@
+             * with internal versioning. It either represents "create" (version
+             * == MATCH_DELETED) or overwrite (version == MATCH_ANY). All other
+             * values are invalid.
+             */
+            if (index.versionType() == INTERNAL) {
+                if (index.version() != Versions.MATCH_DELETED && index.version() != Versions.MATCH_ANY) {
+                    throw new IllegalArgumentException("Invalid version for reindex with internal versioning [" + index.version() + ']');
+                }
+            } else {
+                index.version(doc.version());
+            }
+            return index;
+        }
+
+        /**
+         * Override the simple copy behavior to allow more fine grained control.
+         */
+        @Override
+        protected void copyRouting(IndexRequest index, SearchHit doc) {
+            String routingSpec = mainRequest.destination().routing();
+            if (routingSpec == null) {
+                super.copyRouting(index, doc);
+                return;
+            }
+            if (routingSpec.startsWith("=")) {
+                index.routing(mainRequest.destination().routing().substring(1));
+                return;
+            }
+            switch (routingSpec) {
+            case "keep":
+                super.copyRouting(index, doc);
+                break;
+            case "discard":
+                index.routing(null);
+                break;
+            default:
+                throw new UnsupportedOperationException("Unsupported routing command");
+            }
+        }
+
+        /*
+         * Methods below here handle script updating the index request. They try
+         * to be pretty liberal with regards to types because script are often
+         * dynamically typed.
+         */
+        @Override
+        protected ReindexResponse buildResponse(long took) {
+            return new ReindexResponse(took, created(), updated(), batches(), versionConflicts(), noops(), failures());
+        }
+
+        @Override
+        protected void scriptChangedIndex(IndexRequest index, Object to) {
+            requireNonNull(to, "Can't reindex without a destination index!");
+            index.index(to.toString());
+        }
+
+        @Override
+        protected void scriptChangedType(IndexRequest index, Object to) {
+            requireNonNull(to, "Can't reindex without a destination type!");
+            index.type(to.toString());
+        }
+
+        @Override
+        protected void scriptChangedId(IndexRequest index, Object to) {
+            index.id(Objects.toString(to, null));
+        }
+
+        @Override
+        protected void scriptChangedVersion(IndexRequest index, Object to) {
+            if (to == null) {
+                index.version(Versions.MATCH_ANY).versionType(INTERNAL);
+                return;
+            }
+            index.version(asLong(to, "_version"));
```

Comment:
Do we have some kind of constant class in the code base so we can refer to these standard field by a constant instead of a string literal (sample applies to "_ttl")?